File: insertion_sort.ml
(* Does an insertion sort on `alist` *)
let rec isort alist = 
   match alist with
   | []         -> []
   | head::tail -> insert head (isort tail)

(* a function that inserts item in the proper place *)
and insert item alist =
   match alist with
   | []         -> [item]
   | head::tail ->
      if item <= head then
         item :: alist
      else
         head :: insert item tail (* Note: recursively traverses the list till it reaches a place to insert it *)
;; 

(*

let print_int_space i =
   print_string (string i) ^ " " (* need a way to convert an int to a string*)
;;
*)

let print_int_space i =
   print_int i;
   print_string " ";;

print_string "This is an insertion sort\n";;


print_string "\n from: \t";;

List.map (print_int_space) [9; 4; 6; 7; 5; 1];;


print_string "\n to: \t";;

List.map (print_int_space) (isort [9; 4; 6; 7; 5; 1]);;

File: insertion_sort.ex
defmodule InsertionSort do
  @moduledoc """
  #Insertion_Sort 

  This is writen in Elixir -> http://elixir-lang.org/getting_started/1.html

  Things you need to know:
  - The last statement is automatically returned
  - functions with the same name are the same function they match what's in 
 their argument list


  """


  @doc """
  Will insert an item in the proper part of a list (assuming sorted list)
  """
  def insert(item, []) do
    [item]
  end
  def insert(item, list=[head|tail]) do
    if item <= head do
      [item | list] # finally insert it in proper place
    else
      [head | insert(item, tail)]  # Recursively move it one spot right
    end
  end


  @doc """
  This will do an Insertion Sort
  """
  def sort([]) do
    []
  end
  def sort([head|tail]) do
    insert(head, sort(tail))
  end



  def test(list) do
    IO.puts "Origional list"
    IO.puts Enum.join(list, " ")

    sortedlist = InsertionSort.sort(list)
    IO.puts ""

    IO.puts "Sorted list"
    IO.puts Enum.join(sortedlist, " ")
  end

end


InsertionSort.test([9,4,0,12, 1.9, 1, 12, 134])

